code/modules/events/crystal_event.dm

```diff
@@ -48,9 +47,100 @@ This section is for the event controller
 	var/obj/machinery/destabilized_crystal/dest_crystal
 	///Check if the event will end badly
 	var/is_zk_scenario = TRUE
+	///Ticks that have to pass before second wave
+	var/second_wave = 310
+	///Store the areas where the huge portals appears
+	var/list/center_areas = list()
 
 /datum/round_event/crystal_invasion/start()
+	destabilize_supermatter_crystal()
 	choose_wave_type()
+	addtimer(CALLBACK(src, .proc/create_random_portals), 2 SECONDS)
+
+///Destabilize the Supermatter crystal and spawn 6 portals around it
+/datum/round_event/crystal_invasion/proc/destabilize_supermatter_crystal()
+	var/list/sm_crystal = list()
+	for(var/obj/machinery/power/supermatter_crystal/temp in GLOB.machines)
+		if(istype(temp, /obj/machinery/power/supermatter_crystal/shard))
+			continue
+		sm_crystal += temp
+	if(!length(sm_crystal))
+		log_game("No engine found, killing the crystal invasion event.")
+		kill()
+		return
+
+	var/obj/machinery/power/supermatter_crystal/crystal = pick(sm_crystal)
+	dest_crystal = crystal.destabilize(portal_numbers)
+	RegisterSignal(dest_crystal, COMSIG_PARENT_QDELETING, .proc/on_dest_crystal_qdel)
+
+	for(var/turf/turf_loc in RANGE_TURFS(8, dest_crystal.loc))
+		var/distance_from_center = max(get_dist(turf_loc, dest_crystal.loc), 1)
+		if(prob(325 / distance_from_center))
+			if(isopenturf(turf_loc) || isspaceturf(turf_loc))
+				turf_loc.ChangeTurf(/turf/open/indestructible/crystal_floor, flags = CHANGETURF_INHERIT_AIR)
+			else
+				turf_loc.ChangeTurf(/turf/closed/indestructible/crystal_wall)
+
+	var/list/crystal_spawner_turfs = list()
+	for(var/turf/range_turf in RANGE_TURFS(6, dest_crystal.loc))
+		if(!isopenturf(range_turf) || isspaceturf(range_turf))
+			continue
+		crystal_spawner_turfs += range_turf
+
+	for(var/i in 1 to 6)
+		var/pick_portal = pickweight(GLOB.crystal_invasion_waves["big wave"])
+		var/turf/portal_spawner_turf = pick(crystal_spawner_turfs)
+		new pick_portal(portal_spawner_turf)
+
+///Chooses random areas where to put the main huge portals and then spawn around them 6 more portals
+/datum/round_event/crystal_invasion/proc/create_random_portals()
+	var/list/center_finder = list()
+	for(var/es in GLOB.generic_event_spawns)
+		var/obj/effect/landmark/event_spawn/temp = es
+		if(is_station_level(temp.z))
+			center_finder += temp
+	if(!center_finder.len)
+		kill()
+		CRASH("No landmarks on the station map, aborting")
+	for(var/j in 1 to portal_numbers)
+		var/obj/center = pick(center_finder)
+		var/turf/center_turf = center.loc
+		var/area/center_area = get_area(center_turf)
+		center_areas += center_area
+
+		explosion(center_turf,0,0,5,7,7)
+
+		new/obj/structure/crystal_portal/huge(center_turf)
+
+		for(var/turf/turf_loc in RANGE_TURFS(5, center_turf))
+			var/distance_from_center = max(get_dist(turf_loc, center_turf), 1)
```

Comment:
`get_dist` will only return a non-positive number in case of error. Do you expect this to happen? If so, add a check for it, instead of silently discarding the error.